### PR TITLE
feat: configure and import data from kafka 

### PR DESCRIPTION
**Pull request by elisamozammel**

### CODE REVIEW

File: src/main/java/io/zeebe/monitor/config/KafkaConfig.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import org.springframework.kafka.annotation.KafkaStreamsDefaultConfiguration;
+import org.springframework.kafka.config.KafkaStreamsConfiguration;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Configuration
+@EnableKafka
+@EnableKafkaStreams
+public class KafkaConfig {
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/config/KafkaConfig.java:22]: We aim to use spring cloud streams everywhere across vnoc. Please switch to it.
You can check the config and implementation from vnoc-process-manager for example

File: pom.xml [hunk 2]
```diff
@@ -112,6 +112,22 @@
             <artifactId>spring-boot-starter-websocket</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.springframework.kafka</groupId>
+            <artifactId>spring-kafka</artifactId>
```
Comment (djefimov) [inline pom.xml:120]: what version? We stick to particular kafka version across vnoc

File: src/main/java/io/zeebe/monitor/zeebe/KafkaStreamsService.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    @Autowired
+    private KafkaIncidentImporter kafkaIncidentImporter;
+
+    @Autowired
+    private ZeebeKafkaStreams zeebeKafkaStreams;
+
+    @PostConstruct
+    public void start() {
+        zeebeKafkaStreams.kafkaProcessStream().foreach((key, event) -> {
+            if (event != null) {
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/KafkaStreamsService.java]: can event actually be null?

File: src/main/java/io/zeebe/monitor/zeebe/ZeebeHazelcastService.java [hunk 4]
```diff
@@ -13,7 +13,7 @@
 import javax.annotation.PreDestroy;
 import java.time.Duration;
 
-@Component
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/ZeebeHazelcastService.java:16]: Please make a configurable switch whether to use (enable) hazelcast or kafka

File: src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
+package io.zeebe.monitor.zeebe.kafka;
+
+import java.time.Duration;
+import java.util.Map;
+
+public class GenericRecord {
+    private int partitionId;
+    private Map value;
+    private Long key;
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java]: `Long` -> `long`? ;)

File: src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
+package io.zeebe.monitor.zeebe.kafka;
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java]: should this reside in `dto` package?

File: src/main/java/io/zeebe/monitor/zeebe/kafka/ZeebeKafkaStreams.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
+package io.zeebe.monitor.zeebe.kafka;
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/kafka/ZeebeKafkaStreams.java]: may be put this next to `ZeebeHazelcastService`?

File: src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaErrorImporter.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
+package io.zeebe.monitor.zeebe.kafka.importers;
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaErrorImporter.java]: please don't break package structure 
currently it's `zeebe.importers`
so `zeebe.importers.hazelcast` & `zeebe.importers.kafka` I guess?

File: src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+import java.time.Duration;
+import java.util.Map;
+
+public class GenericRecord {
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java]: `GenericKafkaRecord`?

File: src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaIncidentImporter.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+    final IncidentEntity entity =
+        incidentRepository
+            .findById(key)
+            .orElseGet(
+                () -> {
+                  final IncidentEntity newEntity = new IncidentEntity();
+                  newEntity.setKey(key);
+                  newEntity.setBpmnProcessId((String)values.get("bpmnProcessId"));
+                  newEntity.setProcessDefinitionKey(values.get("processDefinitionKey") != null ? ((Number)values.get("processDefinitionKey")).longValue() : 0);
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaIncidentImporter.java]: hazelcast incident importer (and other importers) doesn't seems to handle `null` for the fields. Why we need `null` handing here?

File: src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaMessageSubscriptionImporter.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  public void importMessageSubscription(final GenericRecord record) {
+
+    final MessageSubscriptionIntent intent =
+        MessageSubscriptionIntent.valueOf(record.getIntent());
+    final long timestamp = record.getTimestamp();
+    Map values = record.getValue();
+    final MessageSubscriptionEntity entity =
+        messageSubscriptionRepository
+            .findByElementInstanceKeyAndMessageName(
+                    values.get("elementInstanceKey") != null ? (Long) values.get("elementInstanceKey") : 0, (String)values.get("messageName"))
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaMessageSubscriptionImporter.java]: you're basically searching by `elementInstanceKey` `0`

File: pom.xml (hunk 2 above)
Comment (jaakkoraut) [reply pom.xml:120]: Library version does not need to be a 100% match by pulling vnoc platform here I think, but it would make sense to specify a current version so that majors match at least

File: src/main/java/io/zeebe/monitor/config/KafkaConfig.java (hunk 1 above)
Comment (jaakkoraut) [reply src/main/java/io/zeebe/monitor/config/KafkaConfig.java:22]: I think so far we only use it in `vnoc-process-manager` while rest of vnoc is using Kafka Streams, I think we should have a discussion with @xulfus on the preferred library

File: src/main/java/io/zeebe/monitor/zeebe/ZeebeHazelcastService.java (hunk 4 above)
Comment (jaakkoraut) [reply src/main/java/io/zeebe/monitor/zeebe/ZeebeHazelcastService.java:16]: It would be beneficial to make this configurable and try to open PR on the upstream project to get more open source community usage of Kafka + Simple Monitor

File: src/main/java/io/zeebe/monitor/config/KafkaConfig.java (hunk 1 above)
Comment (djefimov) [reply src/main/java/io/zeebe/monitor/config/KafkaConfig.java:22]: At least network-actions-gateway and sponde-services are also using it. 
Not to mention that end result looks much cleaner and readable spring cloud config

File: pom.xml (hunk 2 above)
Comment (djefimov) [reply pom.xml:120]: 90% match 😄

File: src/main/java/io/zeebe/monitor/config/KafkaConfig.java (hunk 1 above)
Comment (xulfus) [reply src/main/java/io/zeebe/monitor/config/KafkaConfig.java:22]: Unfortunately we are missing documented guidelines on library choices; my bad, and an action point for me. So, we don't really have an "official" and communicated stand on this. As to spring cloud streams, that is probably a good recommendation - but for this particular case we might decide otherwise as we might want to contribute upstream.

TL;DR for now, I believe Kafka streams is okay for this project as the implementation is already done and there were no other guidelines available

File: src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaErrorImporter.java (hunk 8 above)
Comment (elisamozammel) [reply src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaErrorImporter.java]: My intention was to keep our code separate until get a decision about contribution. Thats why I kept all of our change in one package though kafkaConfig file outside now.

Anyway I can update package to zeebe.importers.kafka but I don't want to touch on existing package for hazelcast otherwise we may face difficulties during getting new changes from zeebe-simple-monitor.

File: src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaIncidentImporter.java (hunk 10 above)
Comment (elisamozammel) [reply src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaIncidentImporter.java]: we don't get data as a same format like hazelcast, We are not able to reuse same entity/dto. we have used generic class where we are getting attributes as a Map.

File: src/main/java/io/zeebe/monitor/zeebe/KafkaStreamsService.java (hunk 3 above)
Comment (elisamozammel) [reply src/main/java/io/zeebe/monitor/zeebe/KafkaStreamsService.java]: Event may not be null but during conversion from json to java object exception may happen

File: pom.xml (hunk 2 above)
Comment (elisamozammel) [reply pom.xml:120]: added specific version

File: src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java (hunk 5 above)
Comment (elisamozammel) [reply src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java]: Fixed

File: src/main/java/io/zeebe/monitor/zeebe/kafka/ZeebeKafkaStreams.java (hunk 7 above)
Comment (elisamozammel) [reply src/main/java/io/zeebe/monitor/zeebe/kafka/ZeebeKafkaStreams.java]: moved

File: src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaErrorImporter.java (hunk 8 above)
Comment (elisamozammel) [reply src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaErrorImporter.java]: refactored

File: src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java (hunk 9 above)
Comment (elisamozammel) [reply src/main/java/io/zeebe/monitor/zeebe/kafka/GenericRecord.java]: refactored

File: src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaMessageSubscriptionImporter.java (hunk 11 above)
Comment (elisamozammel) [reply src/main/java/io/zeebe/monitor/zeebe/kafka/importers/KafkaMessageSubscriptionImporter.java]: if value is then null yes

File: src/main/java/io/zeebe/monitor/config/KafkaConfig.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    @Value("${kafkaBootstrapServers:localhost:9092}")
+    public String kafkaServers;
+    @Value("${kafkaSchemaRegistryUrl:localhost:8081}")
+    public String schemaRegistry;
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/config/KafkaConfig.java]: why we need a schema registry?

File: pom.xml [hunk 13]
```diff
@@ -20,6 +20,8 @@
 
         <zeebe.version>8.1.5</zeebe.version>
         <version.zeebe.spring>8.1.12</version.zeebe.spring>
+        <kafka.version>3.1.2</kafka.version>
```
Comment (LockeBirdsey) [inline pom.xml]: Kafka version should be 3.2.+ due to some weird bugs we've seen in other services. Most other services are using 3.2.+

File: src/main/java/io/zeebe/monitor/config/KafkaConfig.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+    public String getSchemaRegistry() {
+        return schemaRegistry;
+    }
+
+    public int getMaxMessageSizeBytes() {
+        return this.maxMessageSizeBytes;
+    }
+
+    public void setMaxMessageSizeBytes(int var1) {
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/config/KafkaConfig.java]: `var1`?:)

File: src/main/java/io/zeebe/monitor/zeebe/ZeebeKafkaStreams.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ZeebeKafkaStreams {
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    @Autowired
+    public StreamsBuilder kStreamBuilder;
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/ZeebeKafkaStreams.java]: var name is very misleading. You're using `StreamsBuilder` class, but naming it with deprecated `kStreamBuilder`

File: src/main/java/io/zeebe/monitor/rest/dto/GenericKafkaRecord.java [hunk 16]
```diff
@@ -0,0 +1,122 @@
+package io.zeebe.monitor.rest.dto;
+
+import java.util.Map;
+
+public class GenericKafkaRecord {
+    private int partitionId;
+    private Map value;
+    private long key;
+    private Long timestamp;
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/rest/dto/GenericKafkaRecord.java]: Long -> long

File: src/main/java/io/zeebe/monitor/zeebe/ZeebeKafkaStreams.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @Value("${kafka.streams.maxMessageSizeBytes:5242880}")
+    private int maxMessageSizeBytes;
+
+    @Bean
+    public KStream<String, GenericKafkaRecord> kafkaProcessInstanceStream() {
+        KStream stream = this.kStreamBuilder.stream("zeebe-process-instance", Consumed.with(Serdes.String(), Serdes.String())).mapValues (value -> readValue(value));
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/ZeebeKafkaStreams.java]: can we just filter null values inside stream and cut those beforehand?

File: src/main/java/io/zeebe/monitor/zeebe/ZeebeKafkaStreams.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+        KStream stream = this.kStreamBuilder.stream("zeebe-job", Consumed.with(Serdes.String(), Serdes.String())).mapValues (value -> readValue(value));
+        this.logger.debug("Initialize zeebe-job kafka stream");
+        return stream;
+    }
+
+    private GenericKafkaRecord readValue(String value) {
+        try {
+            return objectMapper.readValue(value, GenericKafkaRecord.class);
+        } catch (Exception ex) {
+            ex.printStackTrace();
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/ZeebeKafkaStreams.java:106]: woops

File: src/main/java/io/zeebe/monitor/zeebe/importers/kafka/KafkaJobImporter.java [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+    final long timestamp = record.getTimestamp();
+    Map values = record.getValue();
+    final JobEntity entity =
+        jobRepository
+            .findById(key)
+            .orElseGet(
+                () -> {
+                  final JobEntity newEntity = new JobEntity();
+                  newEntity.setKey(key);
+                  newEntity.setProcessInstanceKey(values.get("processInstanceKey") != null ? ((Number)values.get("processInstanceKey")).longValue() : 0);
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/importers/kafka/KafkaJobImporter.java]: https://stackoverflow.com/a/49894716 - that would make it way shorter

File: src/main/java/io/zeebe/monitor/zeebe/KafkaStreamsService.java [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+        zeebeKafkaStreams.kafkaZeebeMessageSubscriptionStream().foreach((key, event) -> {
+            if (event != null) {
+                kafkaMessageSubscriptionImporter.importMessageSubscription(event);
+            }
+        });
+
+        zeebeKafkaStreams.kafkaZeebeJobStream().foreach((key, event) -> {
+            if (event != null) {
+                kafkaJobImporter.importJob(event);
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/KafkaStreamsService.java]: okay, I played with it a bit
It looks like you don't need own importers a all
You can build a record just by getting values from `GenericKafkaRecord` and mapping it to a required importer record. 
Below sample is based on `JobRecord`:

```
Schema.JobRecord.Builder jobRecordBuilder = Schema.JobRecord.newBuilder();
      jobRecordBuilder.setBpmnProcessId("some value from kafka record");
      //other values
      Schema.JobRecord convertedRecord = jobRecordBuilder.build();
existingJobImporterNotAKafkaSpecificOne.importJob(convertedRecord)
```

Correct me if I'm wrong or it doesn't work

Comment (djefimov) [reply src/main/java/io/zeebe/monitor/zeebe/KafkaStreamsService.java]: https://protobuf.dev/getting-started/javatutorial/#builders-messages - some docs

File: src/main/java/io/zeebe/monitor/zeebe/util/BuildRecordUtil.java [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+import com.google.protobuf.Value;
+import io.zeebe.exporter.proto.Schema;
+import io.zeebe.monitor.rest.dto.GenericKafkaRecord;
+import org.apache.commons.lang3.math.NumberUtils;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Map;
+
+public class BuildRecordUtil {
```
Comment (djefimov) [inline src/main/java/io/zeebe/monitor/zeebe/util/BuildRecordUtil.java:14]: nice! :)

So far it looks like only switch `kafka <-> hazelcast` is left

**Review by djefimov — APPROVED** [review]

File: src/main/resources/application.yaml [hunk 22]
```diff
@@ -59,3 +61,23 @@ white-label:
...
+      on-profile: kafka
+  datasource:
+    url: ${POSTGRES_DATASOURCE_URL:jdbc:postgresql://localhost:5432/postgres}
+    username: ${POSTGRES_DATABASE_USER:postgres}
+    password: ${POSTGRES_DATABASE_PASSWORD:temp}
+    driverClassName: org.postgresql.Driver
+  jpa:
+    database-platform: org.hibernate.dialect.PostgreSQLDialect
+    hibernate:
+      ddl-auto: update
```
Comment (LockeBirdsey) [inline src/main/resources/application.yaml]: ah this might cause some issues with how the DB is configured in the chart already. See https://github.com/elisapolystar/zeebe-simple-monitor-chart/blob/main/templates/deployment.yaml#L53 for how it's done and https://github.com/elisapolystar/deployment-sponde/blob/ca90c6a97f1202a46872656fb8c59b7a6c1f3594/apps/values.yaml#L429 for an example. I would suggest putting these Postgres specific configs in a new profile block or follow how DBs are set in the zeebe-simple-monitor-chart

File: src/main/resources/application.yaml [hunk 23]
```diff
@@ -59,3 +61,23 @@ white-label:
...
+
+kafkaBootstrapServers: ${KAFKA_BOOTSTRAP_SERVERS:localhost:9092}
+kafkaSchemaRegistryUrl: ${SCHEMA_REGISTRY_URL:http://localhost:8081}
+kafka:
+  streams:
+    maxMessageSizeBytes: ${KAFKA_MAX_MESSAGE_SIZE_BYTES:5242880}
```
Comment (LockeBirdsey) [inline src/main/resources/application.yaml:73]: when using this on the cluster, you need to have auth setup (e.g. a `kafka-user` and secrets), this would be done inside the zeebe-simple-monitor-chart repo. Ask your local integrations person for help with that. (I only mentioned it here since Kafka, you don't need any additional changes to the Kafka stuff here)

File: src/main/resources/application.yaml (hunk 22 above)
Comment (elisamozammel) [reply src/main/resources/application.yaml]: Resolved

File: src/main/resources/application.yaml (hunk 23 above)
Comment (elisamozammel) [reply src/main/resources/application.yaml:73]: We will update kafka and deployment related configuration in another PR

**Review by LockeBirdsey — APPROVED** [review]
